In [200]:
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
from sklearn.ensemble import RandomForestClassifier
from sklearn.naive_bayes import GaussianNB
from sklearn.linear_model import LogisticRegression
from sklearn.svm import SVC
from sklearn.metrics import confusion_matrix
from sklearn.model_selection import GridSearchCV
from matplotlib.colors import ListedColormap

In [201]:
# Importing the dataset
dataset = pd.read_csv('dataset.csv') #831x23
#X = dataset.iloc[:, 4:22].values
#y = dataset.iloc[:, 22].values

In [202]:
len(dataset)

831

In [203]:
.8 * 831

664.8000000000001

In [204]:
# Splitting the dataset into the Training set and Test set
#from sklearn.cross_validation import train_test_split
#X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.25, random_state = 0)

X_train = dataset.iloc[0:664, 4:22].values
y_train = dataset.iloc[0:664, 22].values
X_test = dataset.iloc[664:831, 4:22].values
y_test = dataset.iloc[664:831, 22].values

In [205]:
X_train[0]

array([2.50000000e-01, 7.36749117e-01, 7.39045764e-01, 2.25405405e+01,
       1.45961538e+01, 3.26855124e-02, 5.06329114e-02, 1.66067233e+01,
       1.07872257e+01, 7.50000000e-01, 9.47144075e-01, 9.21632653e-01,
       3.26764706e+01, 1.94655172e+01, 2.89855072e-02, 4.73469388e-02,
       3.09493255e+01, 1.79400563e+01])

In [206]:
y_train[0]

0

In [207]:
len(y_train)

664

In [208]:
sc = StandardScaler()
X_train = sc.fit_transform(X_train)
X_test = sc.transform(X_test)


In [209]:
classifier = RandomForestClassifier(n_estimators =10 , criterion = 'entropy', random_state = 0)
classifier.fit(X_train, y_train)

RandomForestClassifier(criterion='entropy', n_estimators=10, random_state=0)

In [210]:
y_pred = classifier.predict(X_test)

In [211]:
cm = confusion_matrix(y_test, y_pred)

In [212]:
cm

array([[21, 38],
       [47, 61]])

In [213]:
classifier.score(X_test,y_test)

0.49101796407185627

In [214]:
y_pred = classifier.predict(X_train)

In [215]:
cm = confusion_matrix(y_train, y_pred)

In [216]:
cm

array([[261,   3],
       [  5, 395]])

In [19]:
n_estimators = [2,4,6,8,10]
min_samples_split = [2,3,4,5,6]
l = [2,3,4,5,6]
criterion = ['entropy','gini']
#param_grid = dict(n_estimators=n_estimators,min_samples_split=min_samples_split,criterion=criterion)

#model = RandomForestClassifier(random_state = 0)
best_test_score=0
best_train_score=0
best_n_estimators=0
min_samples_split=0
criterion=''
diff=100
for n_estimator in [2,4,6,8,10]:
    for split in [2,3,4,5,6]:
        for c in ['entropy','gini']:
            classifier =RandomForestClassifier(n_estimators=n_estimator,min_samples_split=split,criterion=c,random_state = 0)
            classifier.fit(X_train, y_train)
            #y_pred = classifier.predict(X_train)
            train_score = classifier.score(X_train,y_train)
            test_score = classifier.score(X_test,y_test)
            #print("n_estimators = {}, min_samples_split = {}, criterion = {} ,train_score = {},test_score = {}").format(n_estimator,split,c,train_score,test_score)
            #print()
            if(test_score > best_test_score):
                best_test_score=test_score
                best_train_score=train_score
                best_n_estimators=n_estimator
                best_min_samples_split=split
                best_criterion=c
                #diff = (abs(test_score-test_score))
                

print("best_test_score = {},best_train_score = {},best_n_estimators = {} ,min_samples_split = {}, criterion = {}".format(best_test_score,best_train_score,best_n_estimators,best_min_samples_split,best_criterion))        


best_test_score = 0.5688622754491018,best_train_score = 0.9743975903614458,best_n_estimators = 10 ,min_samples_split = 5, criterion = entropy



## best_test_score = 0.918128654971,best_train_score = 0.994293865906,best_n_estimators = 10 ,min_samples_split = 3, criterion = gini

# logistic regression

In [20]:
classifier = LogisticRegression(random_state = 0)
classifier.fit(X_train, y_train)

LogisticRegression(random_state=0)

In [21]:
y_pred = classifier.predict(X_test)
cm = confusion_matrix(y_test, y_pred)
cm

array([[14, 45],
       [13, 95]])

In [218]:
y_pred = classifier.predict(X_train)
cm = confusion_matrix(y_train, y_pred)
cm

array([[ 47, 217],
       [ 31, 369]])

In [219]:

#solver = ['newton-cg', 'lbfgs', 'liblinear', 'sag', 'saga']
#penalty = ['l1', 'l2', 'elasticnet', 'none']
#C = [0.0001,0.001,0.01,0.1,1,10]
#max_iter = [80,90,100]
best_test_score=0
best_train_score=0
best_solver = ''
best_penalty = ''
best_C = 0
best_max_iter = 0
for solver in ['lbfgs', 'liblinear', 'sag', 'saga']:
    for penalty in ['l2']:
        for C in [0.012,0.011,0.0115,0.0116,0.0117,0.0118,0.0119,0.0121,0.0122,0.0123,0.0124]:
            for max_iter in [1,2,3]:
                
                classifier = LogisticRegression(solver=solver,penalty=penalty,C = C,max_iter=max_iter,random_state = 0)
                classifier.fit(X_train, y_train)
                
                train_score = classifier.score(X_train,y_train)
                test_score = classifier.score(X_test,y_test)
                
                if(test_score > best_test_score):
                    best_test_score = test_score
                    best_train_score = train_score
                    best_solver = solver
                    best_penalty = penalty
                    best_C = C
                    best_max_iter = max_iter
                
print("best_test_score = {},best_train_score = {},best_solver = {} ,best_penalty = {}, best_C = {}, best_max_iter = {}".format(best_test_score,best_train_score,best_solver,best_penalty,best_C,best_max_iter))        

                
                

/home/sayak/venv/lib/python3.6/site-packages/sklearn/linear_model/_logistic.py:764: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  extra_warning_msg=_LOGISTIC_SOLVER_CONVERGENCE_MSG)
/home/sayak/venv/lib/python3.6/site-packages/sklearn/linear_model/_logistic.py:764: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic

/home/sayak/venv/lib/python3.6/site-packages/sklearn/linear_model/_logistic.py:764: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  extra_warning_msg=_LOGISTIC_SOLVER_CONVERGENCE_MSG)
/home/sayak/venv/lib/python3.6/site-packages/sklearn/linear_model/_logistic.py:764: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic

/home/sayak/venv/lib/python3.6/site-packages/sklearn/linear_model/_sag.py:330: ConvergenceWarning: The max_iter was reached which means the coef_ did not converge
  "the coef_ did not converge", ConvergenceWarning)
/home/sayak/venv/lib/python3.6/site-packages/sklearn/linear_model/_sag.py:330: ConvergenceWarning: The max_iter was reached which means the coef_ did not converge
  "the coef_ did not converge", ConvergenceWarning)
/home/sayak/venv/lib/python3.6/site-packages/sklearn/linear_model/_sag.py:330: ConvergenceWarning: The max_iter was reached which means the coef_ did not converge
  "the coef_ did not converge", ConvergenceWarning)
/home/sayak/venv/lib/python3.6/site-packages/sklearn/linear_model/_sag.py:330: ConvergenceWarning: The max_iter was reached which means the coef_ did not converge
  "the coef_ did not converge", ConvergenceWarning)
/home/sayak/venv/lib/python3.6/site-packages/sklearn/linear_model/_sag.py:330: ConvergenceWarning: The max_iter was reached which means the 

best_test_score = 0.6706586826347305,best_train_score = 0.6099397590361446,best_solver = sag ,best_penalty = l2, best_C = 0.012, best_max_iter = 1


/home/sayak/venv/lib/python3.6/site-packages/sklearn/linear_model/_sag.py:330: ConvergenceWarning: The max_iter was reached which means the coef_ did not converge
  "the coef_ did not converge", ConvergenceWarning)
/home/sayak/venv/lib/python3.6/site-packages/sklearn/linear_model/_sag.py:330: ConvergenceWarning: The max_iter was reached which means the coef_ did not converge
  "the coef_ did not converge", ConvergenceWarning)
/home/sayak/venv/lib/python3.6/site-packages/sklearn/linear_model/_sag.py:330: ConvergenceWarning: The max_iter was reached which means the coef_ did not converge
  "the coef_ did not converge", ConvergenceWarning)
/home/sayak/venv/lib/python3.6/site-packages/sklearn/linear_model/_sag.py:330: ConvergenceWarning: The max_iter was reached which means the coef_ did not converge
  "the coef_ did not converge", ConvergenceWarning)
/home/sayak/venv/lib/python3.6/site-packages/sklearn/linear_model/_sag.py:330: ConvergenceWarning: The max_iter was reached which means the 

### best_test_score = 0.6706586826347305,best_train_score = 0.6099397590361446,best_solver = sag ,best_penalty = l2, best_C = 0.012, best_max_iter = 1

# Gaussian NB

In [43]:
classifier = GaussianNB()
classifier.fit(X_train, y_train)

train_score = classifier.score(X_train,y_train)
test_score = classifier.score(X_test,y_test)

print("train_score = {} , test_score = {}".format(train_score,test_score))


train_score = 0.6189759036144579 , test_score = 0.5688622754491018


In [44]:
cm = confusion_matrix(y_test, classifier.predict(X_test))
cm

array([[13, 46],
       [26, 82]])

In [45]:
(13.0 + 82)/(13 + 46+26+82)

0.5688622754491018

### train_score = 0.6189759036144579 , test_score = 0.5688622754491018

# SVC

In [49]:
# C = [0.0001,0.001,0.01,0.1,1,10]
# kernel = ['linear', 'poly', 'rbf', 'sigmoid', 'precomputed']
# gamma = ['scale' , 'auto']
# max_iter
best_test_score=0
best_train_score=0
best_C = 0
best_kernel = ''

for C in [0.188,0.189,0.19]:
    for kernel in ['linear', 'poly', 'rbf', 'sigmoid']:
        
        classifier = SVC(C=C,kernel=kernel,random_state = 0)
        classifier.fit(X_train, y_train)
        
        train_score = classifier.score(X_train,y_train)
        test_score = classifier.score(X_test,y_test)

        if(test_score > best_test_score):
            best_C = C
            best_test_score = test_score
            best_train_score = train_score
            best_kernel = kernel
            
print("best_test_score = {},best_train_score = {},best_C = {}, best_kernel = {}".format(best_test_score,best_train_score,best_C,best_kernel))        


best_test_score = 0.6706586826347305,best_train_score = 0.5783132530120482,best_C = 0.189, best_kernel = sigmoid


### best_test_score = 0.6706586826347305,best_train_score = 0.5783132530120482,best_C = 0.189, best_kernel = sigmoid

# ANN

In [136]:
#Importing the Keras libraries and packages
import keras
from keras.models import Sequential
from keras.layers import Dense,Dropout

In [187]:

# Initialising the ANN
classifier = Sequential()

# Adding the input layer and the first hidden layer
# output_dim--no. of node|init--initialises the weights
classifier.add(Dense(32, kernel_initializer="glorot_uniform",bias_initializer="zeros", activation = 'relu'))
classifier.add(Dropout(.2))

classifier.add(Dense(32, kernel_initializer="glorot_uniform",bias_initializer="zeros", activation = 'relu'))
classifier.add(Dropout(.8))

# Adding the second hidden layer
classifier.add(Dense(8, kernel_initializer="glorot_uniform",bias_initializer="zeros", activation = 'relu'))
classifier.add(Dropout(.5))

# Adding the output layer
# sigmoid has been used for binary output.In case of more than two categories we may use softmax activation function
classifier.add(Dense(1,  kernel_initializer="glorot_uniform",bias_initializer="zeros",activation = 'sigmoid'))


In [188]:
opt = keras.optimizers.Adam(learning_rate=0.01)
classifier.compile(optimizer = opt, loss = 'binary_crossentropy',metrics = ['accuracy'])

In [189]:
classifier.fit(X_train, y_train,batch_size=10, epochs=100)

Epoch 1/100
67/67 [==============================] - 0s 3ms/step - loss: 0.8222 - accuracy: 0.5693
Epoch 2/100
67/67 [==============================] - 0s 3ms/step - loss: 0.6890 - accuracy: 0.5979
Epoch 3/100
67/67 [==============================] - 0s 3ms/step - loss: 0.6895 - accuracy: 0.5934
Epoch 4/100
67/67 [==============================] - 0s 3ms/step - loss: 0.6757 - accuracy: 0.5889
Epoch 5/100
67/67 [==============================] - 0s 3ms/step - loss: 0.6709 - accuracy: 0.5994
Epoch 6/100
67/67 [==============================] - 0s 3ms/step - loss: 0.6775 - accuracy: 0.5979
Epoch 7/100
67/67 [==============================] - 0s 3ms/step - loss: 0.6728 - accuracy: 0.6009
Epoch 8/100
67/67 [==============================] - 0s 4ms/step - loss: 0.6788 - accuracy: 0.6024
Epoch 9/100
67/67 [==============================] - 0s 4ms/step - loss: 0.6809 - accuracy: 0.6009
Epoch 10/100
67/67 [==============================] - 0s 3ms/step - loss: 0.6802 - accuracy: 0.5994
Epoch 11/

67/67 [==============================] - 0s 3ms/step - loss: 0.6683 - accuracy: 0.6024
Epoch 84/100
67/67 [==============================] - 0s 3ms/step - loss: 0.6677 - accuracy: 0.6024
Epoch 85/100
67/67 [==============================] - 0s 4ms/step - loss: 0.6715 - accuracy: 0.5994
Epoch 86/100
67/67 [==============================] - 0s 3ms/step - loss: 0.6635 - accuracy: 0.6024
Epoch 87/100
67/67 [==============================] - 0s 3ms/step - loss: 0.6721 - accuracy: 0.6024
Epoch 88/100
67/67 [==============================] - 0s 3ms/step - loss: 0.6650 - accuracy: 0.6024
Epoch 89/100
67/67 [==============================] - 0s 3ms/step - loss: 0.6757 - accuracy: 0.6009
Epoch 90/100
67/67 [==============================] - 0s 3ms/step - loss: 0.6774 - accuracy: 0.6024
Epoch 91/100
67/67 [==============================] - 0s 3ms/step - loss: 0.6732 - accuracy: 0.5994
Epoch 92/100
67/67 [==============================] - 0s 3ms/step - loss: 0.6726 - accuracy: 0.6024
Epoch 93/100


In [190]:
y_pred = classifier.predict(X_test)
y_pred = (y_pred > 0.5)

In [191]:
cm = confusion_matrix(y_test, y_pred)
cm

array([[  0,  59],
       [  0, 108]])

In [186]:
(cm[0][0] + cm[1][1])/(cm[0][0] + cm[0][1] + cm[1][0] + cm[1][1])

0.6467065868263473

In [175]:
scores = classifier.evaluate(X_test, y_test, verbose=0)

In [176]:
print("Accuracy: %.2f%%" % (scores[1]*100))

Accuracy: 61.08%


In [177]:
scores

[0.7270171046257019, 0.6107784509658813]

In [178]:
y_pred = classifier.predict(X_train)
y_pred = (y_pred > 0.5)

In [179]:
cm = confusion_matrix(y_train, y_pred)
cm

array([[ 48, 216],
       [ 12, 388]])

In [180]:
(cm[0][0] + cm[1][1])/(cm[0][0] + cm[0][1] + cm[1][0] + cm[1][1])

0.6566265060240963

### train = 70.2 test= 64.07

# LSTM

In [192]:
from keras.layers import LSTM

In [193]:

X_train, y_train = np.array(X_train), np.array(y_train)
X_train = np.reshape(X_train, (X_train.shape[0], X_train.shape[1], 1))
X_test = np.array(X_test)
X_test = np.reshape(X_test, (X_test.shape[0], X_test.shape[1], 1))

# Initialising the RNN
classifier = Sequential()

# Adding the first LSTM layer and some Dropout regularisation
classifier.add(LSTM(units = 50, return_sequences = True))
classifier.add(Dropout(0.2))

# Adding a second LSTM layer and some Dropout regularisation
classifier.add(LSTM(units = 50, return_sequences = True))
classifier.add(Dropout(0.2))

# Adding a third LSTM layer and some Dropout regularisation
classifier.add(LSTM(units = 50, return_sequences = True))
classifier.add(Dropout(0.2))

# Adding a fourth LSTM layer and some Dropout regularisation
classifier.add(LSTM(units = 50))
classifier.add(Dropout(0.2))

# Adding the output layer
classifier.add(Dense(units = 1))

In [194]:
opt = keras.optimizers.Adam(learning_rate=0.01)
classifier.compile(optimizer = 'adam', loss = 'mean_squared_error')

In [195]:
classifier.fit(X_train, y_train,batch_size=10, epochs=100)

Epoch 1/100
67/67 [==============================] - 4s 57ms/step - loss: 0.2934
Epoch 2/100
67/67 [==============================] - 4s 55ms/step - loss: 0.2497
Epoch 3/100
67/67 [==============================] - 4s 55ms/step - loss: 0.2432
Epoch 4/100
67/67 [==============================] - 4s 56ms/step - loss: 0.2477
Epoch 5/100
67/67 [==============================] - 4s 55ms/step - loss: 0.2478
Epoch 6/100
67/67 [==============================] - 4s 56ms/step - loss: 0.2401
Epoch 7/100
67/67 [==============================] - 4s 56ms/step - loss: 0.2366
Epoch 8/100
67/67 [==============================] - 4s 56ms/step - loss: 0.2353
Epoch 9/100
67/67 [==============================] - 4s 56ms/step - loss: 0.2373
Epoch 10/100
67/67 [==============================] - 4s 56ms/step - loss: 0.2416
Epoch 11/100
67/67 [==============================] - 4s 56ms/step - loss: 0.2387
Epoch 12/100
67/67 [==============================] - 4s 57ms/step - loss: 0.2430
Epoch 13/100
67/67 [=====

In [197]:
y_pred = classifier.predict(X_test)
y_pred = (y_pred > 0.5)

cm = confusion_matrix(y_test, y_pred)
cm

array([[21, 38],
       [41, 67]])

In [198]:
(cm[0][0] + cm[1][1])/(cm[0][0] + cm[0][1] + cm[1][0] + cm[1][1])

0.5269461077844312